src/query/ee/src/fail_safe/handler.rs

```diff
@@ -0,0 +1,384 @@
+            let segment = segment_reader
+                .read(&LoadParams {
+                    location: path.to_string(),
+                    len_hint: None,
+                    ver: *ver,
+                    put_cache: false,
+                })
+                .await;
+
+            match segment {
+                Ok(segment) => {
+                    self.recover_segment(segment.as_ref(), operator.clone())
+                        .await?;
+                }
+                Err(e) => {
+                    if e.code() == ErrorCode::STORAGE_NOT_FOUND {
+                        self.recover_object(path).await?;
+                        // load the segment again
+                        let segment = segment_reader
+                            .read(&LoadParams {
+                                location: path.to_string(),
+                                len_hint: None,
+                                ver: *ver,
+                                put_cache: false,
+                            })
+                            .await?;
+                        self.recover_segment(segment.as_ref(), operator.clone())
+                            .await?;
+                    } else {
+                        info!("other errors (ignored) {}", e);
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    async fn recover_segment(
+        &self,
+        segment: &CompactSegmentInfo,
+        operator: Operator,
+    ) -> Result<()> {
+        let block_metas = segment.block_metas()?;
+        let block_locations = block_metas
+            .iter()
+            .map(|item| &item.location)
+            .collect::<Vec<_>>();
+        self.recover_blocks(&block_locations, operator).await?;
+        Ok(())
+    }
+
+    async fn recover_blocks(
+        &self,
+        snapshot_location: &[&Location],
+        operator: Operator,
+    ) -> Result<()> {
+        for (path, _ver) in snapshot_location {
+            if let Err(e) = operator.stat(path).await {
+                if e.kind() == ErrorKind::NotFound {
+                    self.recover_object(path).await?
+                } else {
+                    return Err(ErrorCode::StorageOther(format!(
+                        "stat block file {} failed {}",
+                        path, e
+                    )));
+                }
+            } else {
+                info!("block at {} is ok", path)
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn recover_object(&self, key: &str) -> Result<()> {
+        let full_path;
+        let obj_key = {
+            if self.root.is_empty() {
+                key
+            } else {
+                full_path = if self.root.ends_with('/') {
+                    format!("{}{}", self.root, key)
+                } else {
+                    format!("{}/{}", self.root, key)
+                };
+                &full_path
+            }
+        };
+
+        // trim the leading "/"
```

Comment:
Why is leading "/" trimmed? (I'm not very familiar with this part)

Reply:
oh, that is a defensive check that prevent "malformed" key prefix passed in from being used as list prefix

Reply:
maybe it is too conservative？